### PR TITLE
Added gMSA accounts support 

### PR DESCRIPTION
**Pull request by anatolybolshakov**
Currently gMSA accounts are not supported by agent.
Recovered [changes](https://github.com/microsoft/azure-pipelines-agent/compare/a5d5e507ef0089917eef73b846372fd990a76a8c..85a5dad321410dafea9d25f5979954cb6a75de0e) made previously.
Issue: https://github.com/microsoft/azure-pipelines-agent/issues/1294
Testing: tested with gmsa account on a local machine with Active Directory account, prepared unit tests for NativeWindowsServiceHelper (IsManagedServiceAccount method)

### CODE REVIEW

**Review by EzzhevNikita — APPROVED** [review]

File: src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs [hunk 1]
```diff
@@ -1256,6 +1323,10 @@ private enum RecoverAction
             RunCommand = 3
         }
 
+        [DllImport("Logoncli.dll", SetLastError = true, CharSet = CharSet.Auto)]
+        private static extern uint NetIsServiceAccount(string ServerName, string AccountName, ref bool IsServiceAccount);
```
Comment (avivanoff) [inline src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs:1327]: [MarshalAs(UnmanagedType.Bool) out bool IsServiceAccount

File: src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs [hunk 2]
```diff
@@ -370,7 +374,7 @@ public bool GrantUserLogonAsServicePrivilage(string domain, string userName)
             }
         }
 
-        public static bool IsWellKnownIdentity(String accountName)
+        public bool IsWellKnownIdentity(String accountName)
         {
             NTAccount ntaccount = new NTAccount(accountName);
```
Comment (avivanoff) [inline src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs:379]: var ntaccount = ...;

File: src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs [hunk 3]
```diff
@@ -370,7 +374,7 @@ public bool GrantUserLogonAsServicePrivilage(string domain, string userName)
...
             SecurityIdentifier sid = (SecurityIdentifier)ntaccount.Translate(typeof(SecurityIdentifier));
```
Comment (avivanoff) [inline src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs:380]: var sid = ...;

File: src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs [hunk 4]
```diff
@@ -923,6 +937,40 @@ public void RevokeDirectoryPermissionForAccount(IList<string> folders)
...
         }
+        /// <summary>
+        /// Checks if account is managed service
+        /// </summary>
+        /// <param name="accountName">account name</param>
+        /// <returns>Returns true if account is managed service.</returns>
+        /// <exception cref="Win32Exception">Throws this exception if there is some error during check.</exception>
+        public bool IsManagedServiceAccount(String accountName)
+        {
+            bool isServiceAccount = false;
```
Comment (avivanoff) [inline src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs:948]: After the change on line 1327, initialization is not needed.

File: src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs [hunk 5]
```diff
@@ -923,6 +937,40 @@ public void RevokeDirectoryPermissionForAccount(IList<string> folders)
...
+            accountName = SanitizeManagedServiceAccountName(accountName);
+            var result = this.CheckNetIsServiceAccount(null, accountName, ref isServiceAccount);
+            if (result == 0)
+            {
+                return isServiceAccount;
+            }
+            else
+            {
+                int lastErrorCode = (int)GetLastError();
```
Comment (avivanoff) [inline src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs:957]: auto lastErrorCode = Marshal.GetLastWin32Error().

**Review by KonstantinTyukalov — APPROVED** [review]

File: src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs (hunk 2 above)
Comment (anatolybolshakov) [reply src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs]: Fixed, thanks!

File: src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs (hunk 3 above)
Comment (anatolybolshakov) [reply src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs]: Fixed

File: src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs [hunk 6]
```diff
@@ -453,6 +457,16 @@ public void InstallService(string serviceName, string serviceDisplayName, string
         {
             Trace.Entering();
 
+            try
+            {
+                var isManagedServiceAccount = IsManagedServiceAccount(logonAccount);
+                Trace.Info($"Account '{logonAccount}' is managed service account: {isManagedServiceAccount}.");
+            }
+            catch (Exception e)
```
[the remark this answers is not in the document]
Comment (anatolybolshakov) [reply src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs]: Fixed

File: src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs (hunk 1 above)
Comment (anatolybolshakov) [reply src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs]: Fixed

File: src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs (hunk 4 above)
Comment (anatolybolshakov) [reply src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs]: Fixed

File: src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs (hunk 5 above)
Comment (anatolybolshakov) [reply src/Agent.Listener/Configuration.Windows/NativeWindowsServiceHelper.cs]: Changed to var, thanks!

**Comment by anatolybolshakov** [comment]
@mmrazik could you please take a look?

**Review by mmrazik — APPROVED** [review]

**Pull request merged**